src/script/lua_api/l_env.cpp

```diff
@@ -624,6 +918,140 @@ int ModApiEnvMod::l_add_entity(lua_State *L)
 	return 1;
 }
 
+// add_3dline(line_params) -> Lua3DLine or nil
+int ModApiEnvMod::l_add_3dline(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+
+	LineParams line_params;
+
+	lua_getfield(L, 1, "type");
+	const char* type = luaL_checkstring(L, -1);
+
+	if (!strcmp(type, "flat"))
+		line_params.type = LineType::FLAT;
+	else if (!strcmp(type, "dihedral"))
+		line_params.type = LineType::DIHEDRAL;
+	else if (!strcmp(type, "sprite"))
+		line_params.type = LineType::SPRITE;
+	else if (!strcmp(type, "cylindric"))
+		line_params.type = LineType::CYLINDRIC;
+	else
+		FATAL_ERROR("Line type can not be empty string!");
+
+	lua_pop(L, 1);
+
+	lua_getfield(L, 1, "start_pos");
+
+	line_params.pos.first = checkFloatPos(L, -1);
+
+	lua_pop(L, 1);
+
+	lua_getfield(L, 1, "end_pos");
+
+	line_params.pos.second = checkFloatPos(L, -1);
+	lua_pop(L, 1);
+
+	lua_getfield(L, 1, "start_color");
+
+	read_color(L, -1, &line_params.color.first);
+	lua_pop(L, 1);
+
+	lua_getfield(L, 1, "end_color");
+
+	read_color(L, -1, &line_params.color.second);
+	lua_pop(L, 1);
+
+	lua_getfield(L, 1, "start_attach_to");
+
+	if (!lua_isnil(L, -1))
+	{
+		ObjectRef* obj_start = checkObject<ObjectRef>(L, -1);
+
+		if (obj_start)
+			line_params.attached_ids.first = ObjectRef::getobject(obj_start)->getId();
+	}
+
+	lua_pop(L, 1);
+
+	lua_getfield(L, 1, "end_attach_to");
+
+	if (!lua_isnil(L, -1))
+	{
+		ObjectRef* obj_end = checkObject<ObjectRef>(L, -1);
+
+		if (obj_end)
+			line_params.attached_ids.second = ObjectRef::getobject(obj_end)->getId();
+	}
+
+	lua_pop(L, 1);
+
+	lua_getfield(L, 1, "width");
+
+	line_params.width = lua_tonumber(L, -1)*BS;
+	lua_pop(L, 1);
+
+	lua_getfield(L, 1, "texture");
+
+	if (!lua_isnil(L, -1))
+		line_params.texture = luaL_checkstring(L, -1);
+
+	lua_pop(L, 1);
```

Comment:
Please shorten the code with `getstringfield_default()` where feasible. Those helper functions were added for repetitive getfield and pop calls.

See `c_converter.h` for similar functions.